tests/autologging/test_autologging_safety_unit.py

```diff
@@ -0,0 +1,827 @@
+        )
+        patch_destination.fn()
+        assert managed_run_mock.call_count == 1
+        assert active_run is not None
+        assert active_run.info.run_id is not None
+
+
+def test_safe_patch_does_not_manage_run_if_unspecified(
+    patch_destination, test_autologging_integration
+):
+
+    active_run = None
+
+    def patch_impl(original, *args, **kwargs):
+        nonlocal active_run
+        active_run = mlflow.active_run()
+        return original(*args, **kwargs)
+
+    with mock.patch(
+        "mlflow.utils.autologging_utils.with_managed_run", wraps=with_managed_run
+    ) as managed_run_mock:
+        safe_patch(
+            test_autologging_integration, patch_destination, "fn", patch_impl, manage_run=False
+        )
+        patch_destination.fn()
+        assert managed_run_mock.call_count == 0
+        assert active_run is None
+
+
+def test_safe_patch_preserves_signature_of_patched_function(
+    patch_destination, test_autologging_integration
+):
+    def original(a, b, c=10, *, d=11):
+        return 10
+
+    patch_destination.fn = original
+
+    patch_impl_called = False
+
+    def patch_impl(original, *args, **kwargs):
+        nonlocal patch_impl_called
+        patch_impl_called = True
+        return original(*args, **kwargs)
+
+    safe_patch(test_autologging_integration, patch_destination, "fn", patch_impl)
+    patch_destination.fn(1, 2)
+    assert patch_impl_called
+    assert inspect.signature(patch_destination.fn) == inspect.signature(original)
+
+
+def test_safe_patch_provides_original_function_with_expected_signature(
+    patch_destination, test_autologging_integration
+):
+    def original(a, b, c=10, *, d=11):
+        return 10
+
+    patch_destination.fn = original
+
+    original_signature = False
+
+    def patch_impl(original, *args, **kwargs):
+        nonlocal original_signature
+        original_signature = inspect.signature(original)
+        return original(*args, **kwargs)
+
+    safe_patch(test_autologging_integration, patch_destination, "fn", patch_impl)
+    patch_destination.fn(1, 2)
+    assert original_signature == inspect.signature(original)
+
+
+def test_exception_safe_function_exhibits_expected_behavior_in_standard_mode():
+    assert not autologging_utils._is_testing()
+
+    @exception_safe_function
+    def non_throwing_function():
+        return 10
+
+    assert non_throwing_function() == 10
+
+    exc_to_throw = Exception("bad implementation")
+
+    @exception_safe_function
+    def throwing_function():
+        raise exc_to_throw
+
+    with mock.patch("mlflow.utils.autologging_utils._logger.warning") as logger_mock:
+        throwing_function()
+        assert logger_mock.call_count == 1
+        message, formatting_arg = logger_mock.call_args[0]
+        assert "unexpected error during autologging" in message
+        assert formatting_arg == exc_to_throw
+
+
+def test_exception_safe_function_exhibits_expected_behavior_in_test_mode(test_mode_on,):
```

Comment:
```suggestion
@pytest.mark.usefixtures(test_mode_on.__name__)
def test_exception_safe_function_exhibits_expected_behavior_in_test_mode():
```

Can we use `pytest.mark.usefixtures` here to avoid adding an unused argument?

Reply:
Absolutely! Thanks for teaching me about this. I've swapped in `usefixtures` in place of unused fixture arguments